pyg_lib/csrc/sampler/cuda/subgraph_kernel.cu

```diff
@@ -0,0 +1,90 @@
+#include <ATen/ATen.h>
+#include <torch/library.h>
+
+#include "pyg_lib/csrc/utils/cuda/helpers.h"
+
+namespace pyg {
+namespace sampler {
+
+namespace {
+
+#define FULL_MASK 0xffffffff
+
+template <typename scalar_t>
+__global__ void subgraph_deg_kernel_impl(
+    const scalar_t* __restrict__ rowptr_data,
+    const scalar_t* __restrict__ col_data,
+    const scalar_t* __restrict__ nodes_data,
+    const scalar_t* __restrict__ to_local_node_data,
+    scalar_t* __restrict__ out_data,
+    int64_t num_nodes) {
+  CUDA_1D_KERNEL_LOOP(scalar_t, thread_idx, 32 * num_nodes) {
+    scalar_t i = thread_idx >> 5;           // thread_idx / 32
+    scalar_t lane = thread_idx & (32 - 1);  // thread_idx % 32
+
+    auto v = nodes_data[i];
+
+    scalar_t deg = 0;
+    for (scalar_t j = rowptr_data[v] + lane; j < rowptr_data[v + 1]; j += 32) {
+      if (to_local_node_data[col_data[j]] >= 0)  // contiguous access
+        deg++;
+    }
+
+    for (scalar_t offset = 16; offset > 0; offset /= 2)  // warp-level reduction
+      deg += __shfl_down_sync(FULL_MASK, deg, offset);
+
+    if (lane == 0)
+      out_data[i] = deg;
+  }
+}
+
+std::tuple<at::Tensor, at::Tensor, c10::optional<at::Tensor>> subgraph_kernel(
+    const at::Tensor& rowptr,
+    const at::Tensor& col,
+    const at::Tensor& nodes,
+    const bool return_edge_id) {
+  TORCH_CHECK(rowptr.is_cuda(), "'rowptr' must be a CUDA tensor");
+  TORCH_CHECK(col.is_cuda(), "'col' must be a CUDA tensor");
+  TORCH_CHECK(nodes.is_cuda(), "'nodes' must be a CUDA tensor");
+
+  const auto stream = at::cuda::getCurrentCUDAStream();
+
+  // We maintain a O(N) vector to map global node indices to local ones.
+  // TODO Can we do this without O(N) storage requirement?
+  const auto to_local_node = nodes.new_full({rowptr.size(0) - 1}, -1);
+  const auto arange = at::arange(nodes.size(0), nodes.options());
+  to_local_node.index_copy_(/*dim=*/0, nodes, arange);
+
+  const auto deg = nodes.new_empty({nodes.size(0)});
+  const auto out_rowptr = rowptr.new_zeros({nodes.size(0) + 1});
+  at::Tensor out_col;
+  c10::optional<at::Tensor> out_edge_id = c10::nullopt;
+
+  AT_DISPATCH_INTEGRAL_TYPES(nodes.scalar_type(), "subgraph_kernel", [&] {
+    const auto rowptr_data = rowptr.data_ptr<scalar_t>();
+    const auto col_data = col.data_ptr<scalar_t>();
+    const auto nodes_data = nodes.data_ptr<scalar_t>();
+    const auto to_local_node_data = to_local_node.data_ptr<scalar_t>();
+    auto deg_data = deg.data_ptr<scalar_t>();
+
+    // Compute induced subgraph degree, parallelize with 32 threads per node:
```

Comment:
I'm actually not sure if it is necessary to parallelize with 32 threads per nodes. Most of the time we are dealing with sparse data and a lot of threads will not go into for loop.

If you are looking for extreme performance, you can bundle `to_local_node_data` and `col_data` into one iterator structure and use this function. I haven't seen any better performance than it in the past.
`https://nvlabs.github.io/cub/structcub_1_1_device_segmented_reduce.html#a4854a13561cb66d46aa617aab16b8825`

Reply:
Do you have an example of bundling `to_local_node_data` and `col_data` into one iterator structure? This looks really interesting.

I am okay with dropping the warp-level parallelism for now, but we will lose the contiguous access to `col_data`, and probably under-utilize the number of threads available on modern GPUs.

Reply:
On a second look, this doesn't seem possible since `col_data` refers to edges, while `to_local_node_data` refers to nodes, while we actually want do the compute across the number of nodes in the induced subgraph.